docs/develop-an-agent.md

```diff
@@ -0,0 +1,350 @@
+auth_type="$(<<< "$identity" jq -r '.spec.auth.type')"
+key_handle="$(<<< "$identity" jq -r '.spec.auth.keyHandle')"
+```
+
+#### **Step 2** Build a bas64-encoded resource URI that expires
+
+Concatenate the URL encoded `resource_uri` with a desired expiry. The script below uses an expiry of 1 day from today. 
+
+In the next step you generate the signed token to connect to IoT Hub by calling the Keys Service's `/sign` API and providing this string as the **message** parameter. The parameter must be a base64-encoded value. 
+
+> **Note**
+>
+> After expiry the client will be unable to connect using the generated SaS token. You will need to repeat the steps to build a new token and client based on a new expiry. This is not shown here.
+
+More details on how the IoT Hub service interprets this signed message can be found in the docs on [Security tokens](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-security#security-tokens). 
+
+```bash
+
+# URL-encode the resource of hub + device + module
+# If we were doing this for the device identity, the resource_uri wouldn't include the '/modules/%s'
+resource_uri="$(printf '%s/devices/%s/modules/%s' "$hub_name" "$device_id" "$module_id" | jq -sRr @uri)"
+
+# Expiry of 1 day from today in Unix epoch time
+expiry="$(bc <<< "$(date +%s) + 60 * 60 * 24")"
+
+# Full string to be signed is a base64 encoded string containing the URL encoded resource URI and the expiry
+signature_data="$(printf '%s\n%s' "$resource_uri" "$expiry" | base64 -w 0)"
+```
+
+#### **Step 3** Get a signed token from the Keys Service
+
+Send a request to the Keys Service endpoint `/run/aziot/keyd.sock` to sign the base64-encoded value using the HMAC-SHA256 algorithm.
+
+If the connection is refused, go back to Step 1 since IoT Hub device information could have changed.
+
+Otherwise, parse the `signature` from the response.
+
+```bash
+# Parse and URL-decode the 'signature' from the response
+signature="$(
+    sudo curl --unix-socket '/run/aziot/keyd.sock' \
+        -X POST -H 'content-type: application/json' --data-raw "$(
+            jq -cn --arg 'keyHandle' "$key_handle" --arg 'message' "$signature_data" '{
+                "keyHandle": $keyHandle,
+                "algorithm": "HMAC-SHA256",
+                "parameters": {
+                    "message": $message,
+                },
+            }'
+        )" 'http://keyd.sock/sign?api-version=2020-09-01' |
+        jq '.signature' -r |
+        sed -e 's|+|%2b|g' -e 's|/|%2f|g' -e 's|=|%3d|g'
+)"
+
+# Construct the full SaS token
+sas_token="$(printf 'sr=%s&se=%s&sig=%s' "$resource_uri" "$expiry" "$signature")"
+```
+
+#### **Step 4**: Create the full connection string
+
+The device may have been configured to connect directly to IoT Hub, or it may be connecting via a gateway. It's direct if the `gatewayHost` obtained previously in **Step 1** is the same as the `hubName`. If they differ then connection string should use the gateway host name.
```

Comment:
```suggestion
The device may have been configured to connect directly to IoT Hub, or it may be connecting via a gateway. It's direct if the `gatewayHost` obtained previously in **Step 1** is the same as the `hubName`. If they differ, then connection string should use the gateway host name.
```